docs/arcadia-users-group/20231104-testing-concepts/lesson.md

```diff
@@ -0,0 +1,572 @@
+E       assert (-9+1.6666666666666667j) == NotImplemented
+
+test_mean.py:34: AssertionError
+===================== 1 failed, 4 passed in 2.71 seconds ======================
+```
+
+In the above case, the pytest package sniffed out the tests in the directory and ran them together to produce a report of the sum of the files and functions matching the regular expression `[Tt]est[-_]*`.
+
+The major benefit a testing framework provides is exactly that, a utility to find and run the tests automatically.
+With pytest, this is the command-line tool called `pytest`.
+When `pytest` is run, it will search all directories below where it was called, find all of the Python files in these directories whose names start or end with `test`, import them, and run all of the functions and classes whose names start with `test` or `Test`.
+This automatic registration of test code saves tons of time and provides a consistent organization framework across Python projects.
+
+When you run `pytest`, it will print a dot (`.`) on the screen for every test that passes, an `F` for every test that fails or where there was an unexpected error.
+In rarer situations you may also see an `s` indicating a skipped tests (because the test is not applicable on your system) or a `x` for a known failure (because the developers could not fix it promptly).
+After the dots, pytest will print summary information.
+
+**Challenge 1: Altering functions to pass all tests**
+
+Without changing the tests, alter the mean.py file from the previous section until it passes.
+When it passes, `pytest` will produce results like the following:
+
+```
+pytest
+```
+
+```
+collected 5 items
+
+test_mean.py .....
+
+========================== 5 passed in 2.68 seconds ===========================
+```
+
+<details>
+ <summary> Challenge solution </summary>
+
+There are many ways this challenge could be solved.
+One way is to check for the presence of complex numbers before calculating the mean.
+
+```
+def mean(num_list):
+    if any(isinstance(num, complex) for num in num_list):
+       return NotImplemented
+    else:
+        try:
+            return sum(num_list)/len(num_list)
+        except ZeroDivisionError :
+            return 0
+        except TypeError as detail :
+            msg = "The algebraic mean of an non-numerical list is undefined.\
+                   Please provide a list of numbers."
+            raise TypeError(detail.__str__() + "\n" +  msg)
+```
+
+</details>
+
+
+Note, using `pytest -v` will result in `pytest` listing which tests are executed and whether they pass or not:
+
+```
+pytest
+```
+
+```
+collected 5 items
+
+test_mean.py .....
+
+test_mean.py::test_ints PASSED
+test_mean.py::test_zero PASSED
+test_mean.py::test_double PASSED
+test_mean.py::test_long PASSED
+test_mean.py::test_complex PASSED
+
+========================== 5 passed in 2.57 seconds ===========================
+```
+
+As we write more code, we would write more tests, and pytest would produce more dots.
+
```

Comment:
```suggestion

### Parametrized tests

Often, different unit tests of the same "unit" have similar logic, but are applied to different inputs; this is the case with our `mean()` test suite above. `pytest` offers a convenient mechanism to write just a single test and apply it to multiple inputs, called test parameterization. This way, adding a new test case is as simple as defining one more input and expected output.

To parametrize a test, we "decorate" it with its expected inputs and outputs, and `pytest` will expand the test to run as many times as needed to check all inputs. In the `pytest` output, this will look exactly like running multiple tests that have different functions.

<details>
<summary> Python decorators </summary>
"Decorators" are a Python way to enhance or wrap a function with additional behavior. They are added just before a function definition, using a name that starts with `@`, and can take arguments separate from the function arguments, for example:

\@log_to_file("output.txt")
def mean(num_list):
  ...

</details>

For example, here's a parametrized version of our test suite for `mean()`. In this case, we decorate the test with two parameters that can be used within the test: `num_list` and `expected_mean`. The first parameter will be used to call the `mean()` function, and the second is its expected result, which the test will validate. We then provide a list of matching pairs of these parameters, each of which will run as a test:

\```
@pytest.mark.parametrize("num_list,exp", [
    ([1,2,3,4,5], 3),
    ([0,2,4,6], 3),
    (range(1,10000), 10000/2.0),
  ])
  def test_mean(num_list, exp):
    obs = mean(num_list)
    assert obs == exp
\```

```